src/core/ChannelHandler.js

```diff
@@ -0,0 +1,103 @@
+import Observable from 'zen-observable';
+
+const events = [
+  'phx_reply',
+  'phx_error',
+  'status_changed',
+  'stopped',
+  'error',
+  'failed',
+  'starting',
+  'started',
+  'deploying',
+  'deployed',
+  'deployment_failed',
+  'ready',
+  'terminated',
+  'snapshot_taken',
+  'snapshot_reverted'
+];
+
+export default class ChannelHandler {
+  constructor(name, socket, globalLogger) {
+    this._name = name;
+    this._channel = socket.channel(this._name);
+    this._stream = this._buildChannelStream(events);
+    this._log = globalLogger.extend(`${this._name}`);
+    this._logger = this._stream.subscribe(value =>
+      this._log(JSON.stringify(value, null, 4))
+    );
+    this.init();
+  }
+
+  _buildChannelStream(eventsList) {
+    return new Observable(stream => {
+      const setChannelEvent = event => {
+        this._channel.on(event, payload => {
+          switch (event) {
+            case 'deployment_failed':
+              stream.error({ event, payload });
+              break;
+            case 'error':
+              stream.error({ event, payload });
+              break;
+            case 'failed':
+              stream.error({ event, payload });
+              break;
+            case 'status_changed':
+              stream.next({
+                event: `status_changed_${payload.data}`,
+                payload
+              });
+              break;
+            default:
+              stream.next({ event, payload });
+          }
+        });
+      };
+
+      eventsList.forEach(event => {
+        setChannelEvent(event);
+      });
+    });
+  }
+
+  on(event, cb) {
+    const ref = this._channel.on(event, data => {
```

Comment:
why isn't `on` using a stream if we're refactoring event listeners into streams?

Reply:
I thought this was decided on the basis of having a one time function rather than having a function which is intended to persist?

Reply:
I also remember that this has to do with the mutation of the `status_changed` events. As `once()` and `sequenceEvents()` hook their promise resolving on the firing of a specific event name, it requires that we mutate the `status_changed` events as the info we are looking for is held within each packets payload:

```javascript
{ // raw data into stream
  "eventName": "status_changed",
  "payload": {
        "data": "terminating"
  }
}

{ // mutated data out of stream
  "eventName": "status_changed_terminating",
  "payload": {
        "data": "terminating"
  }
}
```
Having the `on()` allows access to the raw data.